libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+#include "AP_Beacon_Nooploop.h"
+#include <ctype.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL& hal;
+
+// constructor
+AP_Beacon_Nooploop::AP_Beacon_Nooploop(AP_Beacon &frontend, AP_SerialManager &serial_manager) :
+    AP_Beacon_Backend(frontend)
+{
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon, 0);
+    if (_uart != nullptr) {
+        _uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Beacon, 0));
+    }
+}
+
+// return true if sensor is basically healthy (we are receiving data)
+bool AP_Beacon_Nooploop::healthy()
+{
+    // healthy if we have parsed a message within the past 300ms
+    return ((AP_HAL::millis() - _last_update_ms) < AP_BEACON_TIMEOUT_MS);
+}
+
+// update the state of the sensor
+void AP_Beacon_Nooploop::update(void)
+{
+    // return immediately if not serial port
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // check uart for any incoming messages
+    int16_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        uint8_t b = _uart->read();
+        MsgType type = parse_byte(b);
+        if (type == MsgType::NODE_FRAME2) {
+            //hal.console->printf("nooploop nf2\n");
+            parse_node_frame2();
+        } else if (type == MsgType::SETTING_FRAME0) {
+            //hal.console->printf("nooploop sf0\n");
+            parse_setting_frame0();
+        } 
+    }
+}
+
+void AP_Beacon_Nooploop::request_setting()
+{
+    if (_uart->txspace() >= 128) {
+        uint8_t buf[128] = {0x54, 0, 1};
+        buf[127] = 0x55;
+        _uart->write(buf, 128);
+    }
+}
+
+// process one byte received on serial port
+// message is stored in _msgbuf
+AP_Beacon_Nooploop::MsgType AP_Beacon_Nooploop::parse_byte(uint8_t b)
+{
+    // process byte depending upon current state
+    switch (_state) {
+
```

Comment:
You need to check _msg_len against the size of the array somewhere here to avoid buffer overruns

Reply:
Yes, thank you @andyp1per